mne_bids/write.py

```diff
@@ -1195,12 +1207,14 @@ def write_raw_bids(raw, bids_path, events_data=None,
         raise ValueError('raw_file must be an instance of BaseRaw, '
                          'got %s' % type(raw))
 
-    if not hasattr(raw, 'filenames') or raw.filenames[0] is None:
-        raise ValueError('raw.filenames is missing. Please set raw.filenames'
-                         'as a list with the full path of original raw file.')
+    if raw.preload is not False and not allow_preload:
+        raise ValueError('The data is already loaded from disk and may be '
+                         'altered. See warning for "allow_preload".')
 
-    if raw.preload is not False:
-        raise ValueError('The data should not be preloaded.')
+    if ((not hasattr(raw, 'filenames') or raw.filenames[0] is None) and
+            not allow_preload):
+        raise ValueError('raw.filenames is missing. Please set raw.filenames '
+                         'as a list with the full path of original raw file.')
```

Comment:
That is a good point. I can't think of when non-preloaded raw would be missing `filenames`. However, I am just maintaining the behavior of the existing code, which also checked `filenames` after `raw.preload`. This is tested by running `del raw._filenames` and meddling with the MNE backend, which seems like a highly unrealistic use case.

Reply:
can you git blame to know where this comes from?

Reply:
Looks like `raw.preload` check was added after `raw.filenames` check for different reasons. They do seem redundant now though.

I am not familiar enough with MNE to know if it has any explicit checks that guarantees non-preloaded raws have `filenames`, but we could remove one of them.

Reply:
I decided to just remove the `raw.filenames` check.

In addition to being redundant, the error message doesn't make much sense either because MNE actually doesn't let you change `raw.filenames`. The only way to alter it is through a backend variable `raw._filenames`, which is probably not intended to be used that way.